content/dakshdeephere/threading.md

```diff
@@ -0,0 +1,355 @@
+| 2   | Process switching needs interaction with operating system.        | Thread switching does not need to interact with operating system. |
+| 3 | In multiple processing environments, each process executes the same code but has its own memory and file resources. | All threads can share same set of open files, child processes. |
+| 4 | If one process is blocked, then no other process can execute until the first process is unblocked. | While one thread is blocked and waiting, a second thread in the same task can run. |
+| 5 | Multiple processes without using threads use more resources. | Multiple threaded processes use fewer resources. |
+| 6 | In multiple processes each process operates independently of the others. | One thread can read, write or change another thread's data. |
+
+## Challenges for Programmers while creating Threads
+
+- `Dividing activities` – It involves finding the functions within the job that can be run in parallel on separate processors.
+- `Balance` – The task assigned to each processor must also be equal. Now there can be different parameters for that. One parameter can be, assign equal tasks to each processor. But, tasks assigned to more processor may require higher execution time thus overloading one processor. Thus, simply assigning equal tasks to each processor may not work.
+- `Data splitting` – Another challenge is to split the data required for each task.
+- `Data dependency` – sometimes the data required by one thread (T1) might be produced by another (T2). Thus, T1 can not run before T2. Therefore, it becomes difficult for programmers to code.
+- `Testing and debugging` – Multiple threads running in parallel on multiple cores poses another challenge in the testing of applications.
+
+## Syntax
+
+```c
+#include <pthread.h>
+
+int pthread_create(pthread_t *thread, const pthread_attr_t *attr, void *(*start_routine)(void *), void *arg);
+```
+
+- `thread`: A pointer to a pthread_t variable that will store the thread ID of the newly created thread.
+- `attr`: An optional pointer to a pthread_attr_t structure that specifies the attributes of the thread (such as its stack size or scheduling policy). You can pass NULL for default attributes.
+- `start_routine: A function pointer to the function that will be executed by the thread. It should have the following signature: void *function_name(void *arg). It takes a single argument arg and returns a void * pointer.
+- `arg`: An optional argument that can be passed to the start_routine function. It can be used to provide data or parameters to the thread.
+
+The `pthread_create` function returns 0 on success, indicating that the thread was successfully created. On failure, it returns an error code indicating the cause of the failure.
+
+Remember to link your program with the `pthread` library by adding `-pthread` to the compiler command during the compilation process.
+
+## Examples
```

Comment:
Use meaningful headings which explain the step the learner is going to go into next